File: windows/DMShot/Editor/ViewportMath.cs
using System;
using System.Windows;

namespace DMShot.Editor;

/// Pure zoom/pan geometry for the editor canvas. Stateless. Mirrors the macOS
/// ViewportMath.swift exactly (with mirrored unit tests) — the parity anchor.
public static class ViewportMath
{
    public const double MaxNative = 8.0;
    public const double ZoomStep = 1.15;

    public static double FitScale(Size content, Size viewport, double pad)
    {
        if (content.Width <= 0 || content.Height <= 0) return 1;
        double s = Math.Min((viewport.Width - pad) / content.Width,
                            (viewport.Height - pad) / content.Height);
        return s > 0 ? s : 0.01;
    }

    public static double BaseScale(Size content, Size viewport, double pad)
        => Math.Min(FitScale(content, viewport, pad), 1.0);

    public static double MinScale(Size content, Size viewport, double pad)
        => BaseScale(content, viewport, pad);

    public static double MaxScale(Size content, Size viewport, double pad)
        => Math.Max(BaseScale(content, viewport, pad), MaxNative);

    public static double ClampScale(double s, Size content, Size viewport, double pad)
        => Math.Min(Math.Max(s, MinScale(content, viewport, pad)), MaxScale(content, viewport, pad));

    public static Point Offset(Size content, Size viewport, double scale, Point pan)
    {
        double Axis(double v, double c, double p)
        {
            double scaled = c * scale;
            double centered = (v - scaled) / 2;
            if (scaled <= v) return centered;
            return Math.Min(Math.Max(centered + p, v - scaled), 0);
        }
        return new Point(Axis(viewport.Width, content.Width, pan.X),
                         Axis(viewport.Height, content.Height, pan.Y));
    }

    public static Point ClampPan(Size content, Size viewport, double scale, Point pan)
    {
        var off = Offset(content, viewport, scale, pan);
        double cx = (viewport.Width - content.Width * scale) / 2;
        double cy = (viewport.Height - content.Height * scale) / 2;
        return new Point(off.X - cx, off.Y - cy);
    }

    public static Point ImageToView(Point p, Point origin, double scale, Point offset)
        => new(offset.X + scale * (p.X - origin.X), offset.Y + scale * (p.Y - origin.Y));

    public static Point ViewToImage(Point q, Point origin, double scale, Point offset)
        => new((q.X - offset.X) / scale + origin.X, (q.Y - offset.Y) / scale + origin.Y);

    public static (double Scale, Point Pan) PanForZoomAtPoint(
        Point anchor, Size content, Size viewport, double pad, Point origin,
        double oldScale, Point oldPan, double requestedScale)
    {
        double newScale = ClampScale(requestedScale, content, viewport, pad);
        var oldOffset = Offset(content, viewport, oldScale, oldPan);
        var i = ViewToImage(anchor, origin, oldScale, oldOffset);
        double desiredX = anchor.X - newScale * (i.X - origin.X);
        double desiredY = anchor.Y - newScale * (i.Y - origin.Y);
        double cx = (viewport.Width - content.Width * newScale) / 2;
        double cy = (viewport.Height - content.Height * newScale) / 2;
        return (newScale, new Point(desiredX - cx, desiredY - cy));
    }
}

File: mac/Sources/DMShot/ViewportMath.swift
import CoreGraphics

/// Pure zoom/pan geometry for the editor canvas. Stateless. Mirrored exactly in
/// the Windows port (`ViewportMath.cs`) with mirrored unit tests — this identical
/// math is the macOS/Windows parity anchor.
enum ViewportMath {
    static let maxNative: CGFloat = 8.0
    static let zoomStep: CGFloat = 1.15

    /// Scale at which `content` exactly fits `viewport` minus `pad`.
    static func fitScale(content: CGSize, viewport: CGSize, pad: CGFloat) -> CGFloat {
        guard content.width > 0, content.height > 0 else { return 1 }
        let s = min((viewport.width - pad) / content.width,
                    (viewport.height - pad) / content.height)
        return s > 0 ? s : 0.01
    }

    /// Default display scale: fit, but never upscale a small image past 100%.
    static func baseScale(content: CGSize, viewport: CGSize, pad: CGFloat) -> CGFloat {
        min(fitScale(content: content, viewport: viewport, pad: pad), 1.0)
    }

    static func minScale(content: CGSize, viewport: CGSize, pad: CGFloat) -> CGFloat {
        baseScale(content: content, viewport: viewport, pad: pad)
    }

    static func maxScale(content: CGSize, viewport: CGSize, pad: CGFloat) -> CGFloat {
        max(baseScale(content: content, viewport: viewport, pad: pad), maxNative)
    }

    static func clampScale(_ s: CGFloat, content: CGSize, viewport: CGSize, pad: CGFloat) -> CGFloat {
        min(max(s, minScale(content: content, viewport: viewport, pad: pad)),
            maxScale(content: content, viewport: viewport, pad: pad))
    }

    /// Top-left of the drawn content in view space. Centers each axis when the
    /// content fits; clamps to edges (no gap) when it overflows.
    static func offset(content: CGSize, viewport: CGSize, scale: CGFloat, pan: CGPoint) -> CGPoint {
        func axis(_ v: CGFloat, _ c: CGFloat, _ p: CGFloat) -> CGFloat {
            let scaled = c * scale
            let centered = (v - scaled) / 2
            if scaled <= v { return centered }
            return min(max(centered + p, v - scaled), 0)
        }
        return CGPoint(x: axis(viewport.width, content.width, pan.x),
                       y: axis(viewport.height, content.height, pan.y))
    }

    /// Constrain a pan so it never produces an edge gap (derived from `offset`).
    static func clampPan(content: CGSize, viewport: CGSize, scale: CGFloat, pan: CGPoint) -> CGPoint {
        let off = offset(content: content, viewport: viewport, scale: scale, pan: pan)
        let cx = (viewport.width - content.width * scale) / 2
        let cy = (viewport.height - content.height * scale) / 2
        return CGPoint(x: off.x - cx, y: off.y - cy)
    }

    static func imageToView(_ p: CGPoint, origin: CGPoint, scale: CGFloat, offset: CGPoint) -> CGPoint {
        CGPoint(x: offset.x + scale * (p.x - origin.x),
                y: offset.y + scale * (p.y - origin.y))
    }

    static func viewToImage(_ q: CGPoint, origin: CGPoint, scale: CGFloat, offset: CGPoint) -> CGPoint {
        CGPoint(x: (q.x - offset.x) / scale + origin.x,
                y: (q.y - offset.y) / scale + origin.y)
    }

    /// New (scale, pan) for a zoom that keeps the image point under `anchor` fixed.
    static func panForZoomAtPoint(
        anchor: CGPoint, content: CGSize, viewport: CGSize, pad: CGFloat,
        origin: CGPoint, oldScale: CGFloat, oldPan: CGPoint, requestedScale: CGFloat
    ) -> (scale: CGFloat, pan: CGPoint) {
        let newScale = clampScale(requestedScale, content: content, viewport: viewport, pad: pad)
        let oldOffset = offset(content: content, viewport: viewport, scale: oldScale, pan: oldPan)
        let i = viewToImage(anchor, origin: origin, scale: oldScale, offset: oldOffset)
        let desiredX = anchor.x - newScale * (i.x - origin.x)
        let desiredY = anchor.y - newScale * (i.y - origin.y)
        let cx = (viewport.width - content.width * newScale) / 2
        let cy = (viewport.height - content.height * newScale) / 2
        return (newScale, CGPoint(x: desiredX - cx, y: desiredY - cy))
    }
}
